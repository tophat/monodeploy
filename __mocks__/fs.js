import { fs, vol } from 'memfs'
import path from 'path'

const packagesRoot = path.resolve(process.cwd(), 'packages')
vol.fromJSON(
    {
        'package1/package.json': JSON.stringify({
            "description": "A collection of common visual components.",
            "felibPackageGroup": "components",
            "name": "@thm/fe-common-components",
            "version": "0.1.0",
          }),
        'package2/package.json': '{}',
        'package3/package.json': '{}',
    },
    packagesRoot
)

// This needs to be module.exports, not export default, in order to mock fs properly
module.exports = fs
