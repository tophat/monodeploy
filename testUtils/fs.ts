import { promises as fs } from 'fs'
import { dirname } from 'path'

export async function createFile({
    filePath,
    content,
    cwd,
}: {
    filePath: string
    content?: string
    cwd: string
}): Promise<void> {
    const parent = dirname(filePath)
    const fileContent = content ?? 'some content'
    await fs.mkdir(`${cwd}/${parent}`, { recursive: true })
    await fs.writeFile(`${cwd}/${filePath}`, fileContent)
}
