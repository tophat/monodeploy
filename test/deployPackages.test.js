import path from 'path'
import { vol } from 'memfs'
import { deployPackages as _deployPackages } from '..'
import {
    __setLernaUpdatedSucceeds,
    lernaPublish,
    createGitTag,
    __setLernaPublishSucceeds,
    __setCreateGitTagSucceeds,
} from '../src/command-helpers'
import getPackageInfo from '../src/get-package-info'
import fs from 'fs'

describe('deployPackages function', () => {
    const stdout = jest.fn()
    const stderr = jest.fn()
    const deployPackages = (args = {}) =>
        _deployPackages({ ...args, stderr, stdout })

    beforeEach(async () => {
        stdout.mockClear()
        stderr.mockClear()
        lernaPublish.mockClear()
        createGitTag.mockClear()
    })

    describe('when it succeeds', () => {
        afterEach(() => {
            expect(stdout).toHaveBeenCalledTimes(1)
        })

        it('logs a list of packages', () =>
            expect(deployPackages())
                .resolves.toBeUndefined()
                .then(() => {
                    expect(
                        JSON.parse(stdout.mock.calls[0][0]),
                    ).toMatchSnapshot()
                }))

        it('does the publish if there are packages to publish', () => {
            __setLernaUpdatedSucceeds(true)
            __setLernaPublishSucceeds(true)

            return getPackageInfo().then(packages =>
                deployPackages().then(() => {
                    expect(lernaPublish).toHaveBeenCalled()
                    expect(createGitTag).toHaveBeenCalledTimes(packages.length)
                }),
            )
        })

        it('does not do the publish if there are no packages to publish', () => {
            __setLernaUpdatedSucceeds(true)
            __setLernaPublishSucceeds(true)
            vol.reset()
            vol.fromJSON({ packages: {} }, path.join(process.cwd()))
            return deployPackages().then(() => {
                expect(lernaPublish).not.toHaveBeenCalled()
                expect(createGitTag).not.toHaveBeenCalled()
            })
        })
    })

    const getPackageJsonContents = packageName =>
        fs.readFileSync(
            path.join(
                process.cwd(),
                'packages',
                packageName.replace('@thm/', ''),
                'package.json',
            ),
            'utf8',
        )

    it('updates the package.json files of the sibling packages', async () => {
        await deployPackages()
        const packageInfo = await getPackageInfo()
        packageInfo.forEach(({ name }) => {
            expect(getPackageJsonContents(name)).toMatchSnapshot()
        })
    })

    it('updates the package.json publish config if given', async () => {
        await deployPackages({
            registryUrl: 'http://example.com/production-registry/',
        })
        const packageInfo = await getPackageInfo()
        packageInfo.forEach(({ name }) => {
            expect(getPackageJsonContents(name)).toMatchSnapshot()
        })
    })

    describe('when it fails', () => {
        afterEach(() => {
            expect(stdout).not.toHaveBeenCalled()
        })

        it('throws an error if the lerna publish fails', () => {
            __setLernaUpdatedSucceeds(true)
            __setLernaPublishSucceeds(false)
            return expect(deployPackages()).rejects.toEqual(expect.any(Error))
        })

        it('throws an error if git tagging fails', () => {
            __setLernaUpdatedSucceeds(true)
            __setLernaPublishSucceeds(true)
            __setCreateGitTagSucceeds(false)
            return expect(deployPackages()).rejects.toEqual(expect.any(Error))
        })

        it('does not publish if lerna updated fails', () => {
            __setLernaUpdatedSucceeds(false)
            return expect(deployPackages())
                .rejects.toEqual(expect.any(Error))
                .then(() => {
                    expect(stderr).toHaveBeenCalled()
                    expect(lernaPublish).not.toHaveBeenCalled()
                })
        })
    })
})
