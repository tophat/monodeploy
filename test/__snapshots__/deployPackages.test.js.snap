// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deployPackages function updates the package.json files of the sibling packages 1`] = `
"{
    \\"description\\": \\"A sample package @thm/package1\\",
    \\"felibPackageGroup\\": \\"components\\",
    \\"name\\": \\"@thm/package1\\",
    \\"version\\": \\"1.0.0\\",
    \\"dependencies\\": {
        \\"@thm/package2\\": \\"^0.3.0\\",
        \\"@thm/package3\\": \\"^2.3.1\\",
        \\"lodash\\": \\"*\\"
    },
    \\"publishConfig\\": {}
}"
`;

exports[`deployPackages function updates the package.json files of the sibling packages 2`] = `
"{
    \\"description\\": \\"A sample package @thm/package2\\",
    \\"felibPackageGroup\\": \\"components\\",
    \\"name\\": \\"@thm/package2\\",
    \\"version\\": \\"0.3.0\\",
    \\"dependencies\\": {
        \\"@thm/package3\\": \\"^2.3.1\\"
    },
    \\"publishConfig\\": {}
}"
`;

exports[`deployPackages function updates the package.json files of the sibling packages 3`] = `
"{
    \\"description\\": \\"A sample package @thm/package3\\",
    \\"felibPackageGroup\\": \\"components\\",
    \\"name\\": \\"@thm/package3\\",
    \\"version\\": \\"2.3.1\\",
    \\"dependencies\\": {
        \\"dayum\\": \\"*\\"
    },
    \\"publishConfig\\": {}
}"
`;

exports[`deployPackages function updates the package.json publish config if given 1`] = `
"{
    \\"description\\": \\"A sample package @thm/package1\\",
    \\"felibPackageGroup\\": \\"components\\",
    \\"name\\": \\"@thm/package1\\",
    \\"version\\": \\"1.0.0\\",
    \\"dependencies\\": {
        \\"@thm/package2\\": \\"^0.3.0\\",
        \\"@thm/package3\\": \\"^2.3.1\\",
        \\"lodash\\": \\"*\\"
    },
    \\"publishConfig\\": {
        \\"registry\\": \\"http://example.com/production-registry/\\"
    }
}"
`;

exports[`deployPackages function updates the package.json publish config if given 2`] = `
"{
    \\"description\\": \\"A sample package @thm/package2\\",
    \\"felibPackageGroup\\": \\"components\\",
    \\"name\\": \\"@thm/package2\\",
    \\"version\\": \\"0.3.0\\",
    \\"dependencies\\": {
        \\"@thm/package3\\": \\"^2.3.1\\"
    },
    \\"publishConfig\\": {
        \\"registry\\": \\"http://example.com/production-registry/\\"
    }
}"
`;

exports[`deployPackages function updates the package.json publish config if given 3`] = `
"{
    \\"description\\": \\"A sample package @thm/package3\\",
    \\"felibPackageGroup\\": \\"components\\",
    \\"name\\": \\"@thm/package3\\",
    \\"version\\": \\"2.3.1\\",
    \\"dependencies\\": {
        \\"dayum\\": \\"*\\"
    },
    \\"publishConfig\\": {
        \\"registry\\": \\"http://example.com/production-registry/\\"
    }
}"
`;

exports[`deployPackages function when it succeeds logs a list of packages 1`] = `
Array [
  Object {
    "description": "A sample package @thm/package1",
    "group": "components",
    "name": "@thm/package1",
    "version": "1.0.0",
  },
  Object {
    "description": "A sample package @thm/package2",
    "group": "components",
    "name": "@thm/package2",
    "version": "0.3.0",
  },
  Object {
    "description": "A sample package @thm/package3",
    "group": "components",
    "name": "@thm/package3",
    "version": "2.3.1",
  },
]
`;
